### PR TITLE
devengage-1574-implementing-code-fences

### PR DESCRIPTION
**Pull request by Zino-ctrlZ**

### CODE REVIEW

File: app/src/components/formdemo/FormDemo.tsx [hunk 1]
```diff
@@ -749,6 +750,35 @@ export default function FormDemo() {
 		),
 	});
 
+	const yuri_data = `{
```
Comment (timsmithgenesys) [inline app/src/components/formdemo/FormDemo.tsx]: The object depicted in the string is actually a JavaScript object, not a JSON object. JSON objects _must_ be formatted using double quotes like this:

```json
{
  "author": "Yuri the Yeti",
  "gender": "Male",
  "age": "33",
  "hobbies": [
    "sledding",
    "ice fishing",
    "building snow man"
  ],
  "children": []
}
```

This distinction is important because JavaScript apps will automatically parse JSON objects as JavaScript objects, but JSON parsers are not able to parse a JavaScript object as JSON.

File: app/src/components/formdemo/FormDemo.tsx [hunk 2]
```diff
@@ -749,6 +750,35 @@ export default function FormDemo() {
...
+						{`<CodeFence 
+	title="Yuri data in json"
+	autoCollapse={false}
+	value={yuri_data}
+	language="json"
+/>`}
+					</code>
+				</pre>
+				<CodeFence title="Yuri data in json" autoCollapse={false} value={yuri_data} language="json" />
+				<CodeFence title="Yuri data in java" autoCollapse={false} value={yuri_data} language="java" />
```
Comment (timsmithgenesys) [inline app/src/components/formdemo/FormDemo.tsx]: Use a separate value that's actually Java code (and C code for the next one, or change/delete it if you don't want to write C). It doesn't have to be much, but it needs to show the syntax highlighter is working by successfully parsing the value for that language.

File: rollup.config.js [hunk 3]
```diff
@@ -3,17 +3,18 @@ import resolve from '@rollup/plugin-node-resolve';
...
+import json from '@rollup/plugin-json';
 
 const packageJson = require('./package.json');
 
 export default {
 	input: packageJson.exports.require,
 	output: [
 		{
-			file: packageJson.exports.default,
+			dir: packageJson.exports.default,
```
Comment (timsmithgenesys) [inline rollup.config.js]: Why did you change this?

File: rollup.config.js [hunk 4]
```diff
@@ -3,17 +3,18 @@ import resolve from '@rollup/plugin-node-resolve';
...
 			format: 'esm',
 			sourcemap: true,
 		},
 	],
-	plugins: [peerDepsExternal(), resolve(), commonjs(), typescript({ useTsconfigDeclarationDir: true }), postcss()],
+	plugins: [peerDepsExternal(), resolve(), commonjs(), typescript({ useTsconfigDeclarationDir: true }), json(), postcss()],
```
Comment (timsmithgenesys) [inline rollup.config.js]: Why is this plugin needed? I don't see any JSON files added to the source in this PR.

File: build/index.js/index.js [hunk 5]
```diff
@@ -0,0 +1,2232 @@
+import { GenesysDevIcon, GenesysDevIcons } from 'genesys-dev-icons';
```
Comment (timsmithgenesys) [inline build/index.js/index.js]: this file doesn't appear to be in the right place

File: src/codefence/CodeFence.scss [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+				padding-top: 3px;
+			}
+		}
+
+		pre {
+			display: block;
+			margin: 0 !important;
+			padding: 0 !important;
+			border-radius: 0 0 4px 4px !important;
+			background: var($--theme-codefence-background-color) !important;
```
Comment (timsmithgenesys) [inline src/codefence/CodeFence.scss:78]: Try removing all of these `!important` tags and see if the style looks correct without them. If any of them are doing anything, see if you can fix it without using that tag. There's a good chance they're not needed anymore; this is likely leftover from when the UI was based on bootstrap and this needed some force to override those styles.

**Review by timsmithgenesys — CHANGES_REQUESTED** [review]
There are a few issues that need to be addressed. Let me know if you have questions about any of it. 

Don't forget to bump the version number to `0.3.3`

File: rollup.config.js (hunk 4 above)
Comment (Zino-ctrlZ) [reply rollup.config.js]: This plugin was needed because of dependencies in CodeFence, I believe its the `react-syntax-highlighter` dependency. Without the plugin, there's an error saying the plugin needs to be installed to import json.

File: rollup.config.js (hunk 3 above)
Comment (Zino-ctrlZ) [reply rollup.config.js]: it was in response to an error with this message `Error: When building multiple chunks, the output.dir option must be used, not output.file.`     Output.dir is used when multiple javascript chunks are generated  in  preference to output.file when just one javascript chunk is generated. https://rollupjs.org/guide/en/#outputdir

Comment (timsmithgenesys) [reply rollup.config.js]: This seems suspicious. The structure of the package shouldn't have changed. Maybe this is related to the deleted build artifacts and new incorrectly located file `build/index.js/index.js`.

File: rollup.config.js (hunk 4 above)
Comment (timsmithgenesys) [reply rollup.config.js]: Can you format the plugins array to be one on each line and add a comment for this plugin to explain that it's required for that package?

File: src/codefence/CodeFence.scss (hunk 6 above)
Comment (Zino-ctrlZ) [reply src/codefence/CodeFence.scss:78]: Was successful removing some but , there are some elements whose attributes are set inline style by SyntaxHighlighter, so `!Important` is necessary is we want our styling to persist

File: rollup.config.js (hunk 3 above)
Comment (Zino-ctrlZ) [reply rollup.config.js]: good point, but I updated the file location and reverted to using output.file and it produced the same error `[!] Error: Invalid value for option "output.file" - when building multiple chunks, the "output.dir" option must be used, not "output.file". To inline dynamic imports, set the "inlineDynamicImports" option.`

File: rollup.config.js (hunk 4 above)
Comment (Zino-ctrlZ) [reply rollup.config.js]: got it

File: app/package.json [hunk 7]
```diff
@@ -5,7 +5,6 @@
 	"homepage": "https://purecloudlabs.github.io/genesys-react-components",
 	"dependencies": {
 		"genesys-dev-icons": "^0.1.30",
-		"genesys-react-components": "^0.3.2",
```
Comment (timsmithgenesys) [inline app/package.json:8]: Make sure you run `npm run unlink` in this dir to revert the local link and add the normal dependency back to the config. This leaves the project in a usable state for other users that may not have the local link set up.

File: rollup.config.js [hunk 8]
```diff
@@ -3,17 +3,23 @@ import resolve from '@rollup/plugin-node-resolve';
...
 			format: 'esm',
 			sourcemap: true,
 		},
 	],
-	plugins: [peerDepsExternal(), resolve(), commonjs(), typescript({ useTsconfigDeclarationDir: true }), postcss()],
+	plugins: [peerDepsExternal(),
+		resolve(),
+		commonjs(),
+		typescript({ useTsconfigDeclarationDir: true }),
+		json(),
```
Comment (timsmithgenesys) [inline rollup.config.js]: Don't forget to comment this one about what purpose it serves. It's not obvious because the project neither directly contains nor directly outputs any JSON files.

**Review by timsmithgenesys — CHANGES_REQUESTED** [review]

**Comment by timsmithgenesys** [comment]
These files still aren't in the right place. Also, what's the purpose of the one with some generated string in the filename?

image

File: rollup.config.js (hunk 8 above)
Comment (Zino-ctrlZ) [reply rollup.config.js]: I commented on the line where the file was imported

**Review by timsmithgenesys — CHANGES_REQUESTED** [review]
Still need to fix the file locations. `build/index.js/index.js` is not correct (nor any of the other paths in the screenshot).

image

File: package.json [hunk 9]
```diff
@@ -43,12 +43,12 @@
 		"uuid": "^9.0.0"
 	},
 	"peerDependencies": {
+		"@rollup/plugin-json": "^6.0.0",
```
Comment (timsmithgenesys) [inline package.json]: this plugin isn't needed anymore since react-syntax-highlighter isn't being compiled into the package anymore

File: rollup.config.js [hunk 10]
```diff
@@ -3,6 +3,7 @@ import resolve from '@rollup/plugin-node-resolve';
 import commonjs from '@rollup/plugin-commonjs';
 import typescript from 'rollup-plugin-typescript2';
 import postcss from 'rollup-plugin-postcss';
+import json from '@rollup/plugin-json'; //required for react-syntax-highlighter package in CodeFence.tsx
```
Comment (timsmithgenesys) [inline rollup.config.js]: and remove the package's usages

**Review by timsmithgenesys — CHANGES_REQUESTED** [review]
clean up the unused package then do a branch build to verify functionality in the published package

**Review by timsmithgenesys — APPROVED** [review]
